charms/kfp-persistence/src/components/sa_token_component.py

```diff
@@ -0,0 +1,105 @@
+        **kwargs,
+    ):
+        """Instantiate the SaTokenComponent.
+
+        Args:
+            audiences (List[str]): list of audiences for the SA token
+            expiration (int): token expiration time in seconds
+            filename (str): filename to save a token file
+            path (str): path to save a token file
+            sa_name (str): ServiceAccount name
+            sa_namespace (str): ServiceAccount namespace
+        """
+        super().__init__(*args, **kwargs)
+        self._audiences = audiences
+        self._expiration = expiration
+        self._filename = filename
+        self._sa_name = sa_name
+        self._sa_namespace = sa_namespace
+        self._path = path
+
+    @property
+    def kubernetes_client(self) -> CoreV1Api:
+        """Load cluster configuration and return a CoreV1 Kubernetes client."""
+        try:
+            kubernetes.config.load_incluster_config()
+        except ConfigException:
+            kubernetes.config.load_kube_config()
+
+        api_client = kubernetes.client.ApiClient()
+        core_v1_api = kubernetes.client.CoreV1Api(api_client)
+        return core_v1_api
+
+    def _create_sa_token(self) -> AuthenticationV1TokenRequest:
+        """Return a TokenRequest."""
+        # The TokenRequest should always have the audience pointing to pipelines.kubeflow.org
+        # and an large expiration time to avoid having to re-generate the token and push it
+        # again to the workload container.
+        spec = V1TokenRequestSpec(audiences=self._audiences, expiration_seconds=self._expiration)
+        body = kubernetes.client.AuthenticationV1TokenRequest(spec=spec)
+        try:
+            api_response = self.kubernetes_client.create_namespaced_service_account_token(
+                name=self._sa_name, namespace=self._sa_namespace, body=body
+            )
+        except ApiException as e:
+            logger.error("Error creating the sa token.")
+            raise e
+        return api_response
+
+    def _generate_and_save_token(self, path: str, filename: str) -> None:
+        """Save the sa token in path in the charm container.
+
+        Args:
+            path (str): a path to store the token file
+            filename (str): a filename for the token file
+        """
+        if not Path(path).is_dir():
+            logger.error("Path does not exist, cannot proceed saving the sa token file.")
+            raise RuntimeError("Path does not exist, cannot proceed saving the sa token file.")
+        if Path(path, filename).is_file():
+            logger.info("Token file already exists, nothing else to do.")
+        api_response = self._create_sa_token()
+        token = api_response.status.token
+        with open(Path(path, filename), "w") as token_file:
+            token_file.write(token)
+
+    def get_status(self) -> StatusBase:
+        """Return the status of this component."""
+        try:
+            self._generate_and_save_token(self._path, self._filename)
```

Comment:
Is this how `get_status()` works? I wonder what will happen if `_generate_and_save_token()` raises an error. Shouldn't `_generate_and_save_token()` return a value to update status accordingly? (more of a charm reconciler question)

Reply:
I did a small refactor to make this easier to read and understand, as well as improve the utilisation of the base charm. I am now generating the sa token file inside the `_configure_app_leader` method in the sa token component (according to the [guide](https://github.com/canonical/charmed-kubeflow-chisme/blob/main/src/charmed_kubeflow_chisme/components/how-to/authoring-a-simple-component.md#defining-what-a-component-does-during-charm-reconciliation)).
Please note I also modified the `get_status()` method so it returns active in the presence of the sa token file, and raises a `GenericCharmRuntime` exception otherwise. About the latter, the base charm currently sets the unit status arbitrarily in the presence of an exception, which we should change in the future.

Reply:
Great addition, thank you @DnPlas. Could we have an issue for this in chisme repo? and we can discuss after release once Andrew is also back.